File: python/image_stream.py
import sys
import re
import struct
import base64
import threading
import socket
import configparser

import numpy as np
import cv2 as cv

import server_utils

class ImageStream:
    def get_frame(self):
        raise NotImplementedError()

class CameraImageStream(ImageStream):
    def __init__(self, url, scale):
        if not url.startswith('http'):
            url = int(url)
        self.cap = cv.VideoCapture(url)
        if self.cap.isOpened():
            self.cap.set(cv.CAP_PROP_FRAME_WIDTH, self.cap.get(cv.CAP_PROP_FRAME_WIDTH) * scale)
            self.cap.set(cv.CAP_PROP_FRAME_HEIGHT, self.cap.get(cv.CAP_PROP_FRAME_HEIGHT) * scale)

    def close(self):
        if self.cap.isOpened():
            self.cap.release()

    def get_frame(self):
        success, frame = self.cap.read()
        return frame

class ThreadedImageStream(ImageStream):
    def __init__(self, buffer_size):
        self.ring_buffer = [None for i in range(buffer_size)]
        self.size = 0
        self.wptr = 0
        self.rptr = 0
        self.worker_thread = None
        self.lock = threading.Lock()
        self.cv = threading.Condition(self.lock)
        self.running = False

    def close(self):
        with self.lock:
            was_running = self.running
            self.running = False
        if was_running:
            self.worker_thread.join()
            self.worker_thread = None

    def get_frame(self):
        with self.cv:
            self.cv.wait_for(lambda: not self.running or self.size)
            frame = None
            if self.size:
                frame = self.ring_buffer[self.rptr]
                self.rptr = (self.rptr + 1) % len(self.ring_buffer)
                self.size -= 1
        return frame

    def start(self):
        self.running = True
        self.worker_thread = threading.Thread(target=self.worker)
        self.worker_thread.start()

    def fetch_data(self):
        raise NotImplementedError()

    def worker(self):
        while True:
            with self.lock:
                if not self.running:
                    break
            data = self.fetch_data()
            frame = None
            if data:
                frame = cv.imdecode(np.frombuffer(data, dtype=np.uint8), cv.IMREAD_COLOR)
            with self.cv:
                self.ring_buffer[self.wptr] = frame
                self.wptr = (self.wptr + 1) % len(self.ring_buffer)
                if self.size == len(self.ring_buffer):
                    self.rptr = (self.rptr + 1) % len(self.ring_buffer)
                else:
                    self.size += 1
                self.cv.notify(1)

class PipeImageStream(ThreadedImageStream):
    def __init__(self, buffer_size):
        super().__init__(buffer_size)

        self.start()

    def fetch_data(self):
        try:
            length_bytes = bytearray()
            while len(length_bytes) < 8:
                length_bytes += sys.stdin.read(8 - len(length_bytes))
            length = struct.unpack('<Q', length_bytes)[0]
            data = bytearray()
            while len(data) < length:
                data += sys.stdin.read(length - len(data))
            return base64.b64decode(data)
        except Exception:
            return b''

class SocketImageStream(ThreadedImageStream):
    def __init__(self, addr, port, buffer_size):
        super().__init__(buffer_size)

        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            self.sock.connect((addr, port))
            self.start()
        except Exception:
            pass

    def fetch_data(self):
        try:
            length_bytes = bytearray()
            while len(length_bytes) < 8:
                length_bytes += self.sock.recv(8 - len(length_bytes))
            length = struct.unpack('<Q', length_bytes)[0]
            data = bytearray()
            while len(data) < length:
                data += self.sock.recv(length - len(data))
            return base64.b64decode(data)
        except Exception:
            return b''

def create_image_stream():
    config = configparser.ConfigParser()
    config.read('image_stream.ini')
    stream_type = config.get('DEFAULT', 'stream_type', fallback='camera')
    camera_url = config.get('DEFAULT', 'camera_url', fallback=0)
    scale = config.getint('DEFAULT', 'scale', fallback=1)
    buffer_size = config.getint('DEFAULT', 'buffer_size', fallback=64)
    server = config.get('DEFAULT', 'server', fallback='127.0.0.1:80')
    ip, port = server_utils.parse_server_addr(server)
    if stream_type == 'camera':
        image_stream = CameraImageStream(camera_url, scale)
    elif stream_type == 'pipe':
        image_stream = PipeImageStream(buffer_size)
    elif stream_type == 'socket':
        image_stream = SocketImageStream(ip, port, buffer_size)
    return image_stream

File: src/image_codec/image_stream.cpp
#include <iostream>
#include <fstream>

#include <boost/program_options.hpp>

#include "image_stream.h"
#include "base64.h"
#include "server_utils.h"

CameraImageStream::CameraImageStream(const std::string& url, float scale, int width, int height) {
    if (url.find("http") == 0) {
        m_cap.open(url);
    } else {
        m_cap.open(std::stoi(url));
    }
    if (m_cap.isOpened()) {
        if (scale) {
            m_cap.set(cv::CAP_PROP_FRAME_WIDTH, m_cap.get(cv::CAP_PROP_FRAME_WIDTH) * scale);
            m_cap.set(cv::CAP_PROP_FRAME_HEIGHT, m_cap.get(cv::CAP_PROP_FRAME_HEIGHT) * scale);
        } else {
            m_cap.set(cv::CAP_PROP_FRAME_WIDTH, width);
            m_cap.set(cv::CAP_PROP_FRAME_HEIGHT, height);
        }
    }
}

CameraImageStream::~CameraImageStream() {
    if (m_cap.isOpened()) {
        m_cap.release();
    }
}

cv::Mat CameraImageStream::GetFrame() {
    cv::Mat frame;
    m_cap >> frame;
    return frame;
}

ThreadedImageStream::~ThreadedImageStream() {
    if (m_runnning) {
        m_runnning = false;
        m_thread.join();
    }
}

cv::Mat ThreadedImageStream::GetFrame() {
    std::unique_lock<std::mutex> lock(m_mtx);
    m_cv.wait(lock, [this] { return !m_runnning || m_size; });
    cv::Mat frame;
    if (m_size) {
        frame = std::move(m_ring_buffer[m_rptr]);
        IncPtr(m_rptr);
        --m_size;
    }
    return frame;
}

void ThreadedImageStream::Start() {
    m_runnning = true;
    m_thread = std::thread(&ThreadedImageStream::Worker, this);
}

void ThreadedImageStream::Worker() {
    while (m_runnning) {
        Bytes data = FetchData();
        cv::Mat frame;
        if (!data.empty()) {
            frame = cv::imdecode(data, cv::IMREAD_COLOR);
        }
        {
            std::lock_guard<std::mutex> lock(m_mtx);
            m_ring_buffer[m_wptr] = std::move(frame);
            IncPtr(m_wptr);
            if (m_size == m_ring_buffer.size()) {
                IncPtr(m_rptr);
            } else {
                ++m_size;
            }
            m_cv.notify_one();
        }
    }
}

void ThreadedImageStream::IncPtr(size_t& ptr) {
    ptr = (ptr + 1) % m_ring_buffer.size();
}

PipeImageStream::PipeImageStream(size_t buffer_size) : ThreadedImageStream(buffer_size) {
    Start();
}

Bytes PipeImageStream::FetchData() {
    try {
        uint64_t len = 0;
        uint64_t done1 = 0;
        while (done1 < sizeof(uint64_t)) {
            std::cin.read(reinterpret_cast<char*>(&len) + done1, sizeof(uint64_t) - done1);
            done1 += std::cin.gcount();
        }
        Bytes data(len);
        uint64_t done2 = 0;
        while (done2 < len) {
            std::cin.read(reinterpret_cast<char*>(data.data()) + done2, len - done2);
            done2 += std::cin.gcount();
        }
        return b64decode(data);
    } catch (std::exception& e) {
        return Bytes();
    }
}

SocketImageStream::SocketImageStream(const std::string& addr, int port, size_t buffer_size) : ThreadedImageStream(buffer_size) {
    boost::asio::ip::tcp::endpoint endpoint(boost::asio::ip::make_address(addr), port);
    try {
        m_socket.connect(endpoint);
        Start();
    } catch (std::exception& e) {
    }
}

Bytes SocketImageStream::FetchData() {
    try {
        uint64_t len = 0;
        uint64_t done1 = 0;
        while (done1 < sizeof(uint64_t)) {
            done1 += boost::asio::read(m_socket, boost::asio::buffer(reinterpret_cast<char*>(&len) + done1, sizeof(uint64_t) - done1));
        }
        Bytes data(len);
        uint64_t done2 = 0;
        while (done2 < len) {
            done2 += boost::asio::read(m_socket, boost::asio::buffer(data.data() + done2, len - done2));
        }
        return b64decode(data);
    } catch (std::exception& e) {
        return Bytes();
    }
}

std::unique_ptr<ImageStream> create_image_stream() {
    std::string stream_type("camera");
    std::string camera_url("0");
    float scale = 1;
    int width = 800;
    int height = 600;
    size_t buffer_size = 64;
    std::string server("127.0.0.1:80");
    std::ifstream cfg_file("image_stream.ini");
    boost::program_options::options_description desc;
    auto desc_handler = desc.add_options();
    desc_handler("DEFAULT.stream_type", boost::program_options::value<std::string>(&stream_type));
    desc_handler("DEFAULT.camera_url", boost::program_options::value<std::string>(&camera_url));
    desc_handler("DEFAULT.scale", boost::program_options::value<float>(&scale));
    desc_handler("DEFAULT.width", boost::program_options::value<int>(&width));
    desc_handler("DEFAULT.height", boost::program_options::value<int>(&height));
    desc_handler("DEFAULT.buffer_size", boost::program_options::value<size_t>(&buffer_size));
    desc_handler("DEFAULT.server", boost::program_options::value<std::string>(&server));
    boost::program_options::variables_map vm;
    store(parse_config_file(cfg_file, desc, false), vm);
    notify(vm);
    auto [ip, port] = parse_server_addr(server);
    std::unique_ptr<ImageStream> image_stream;
    if (stream_type == "camera") {
        image_stream = std::make_unique<CameraImageStream>(camera_url, scale, width, height);
    } else if (stream_type == "pipe") {
        image_stream = std::make_unique<PipeImageStream>(buffer_size);
    } else if (stream_type == "socket") {
        image_stream = std::make_unique<SocketImageStream>(ip, port, buffer_size);
    }
    return image_stream;
}
